pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go

```diff
@@ -491,6 +494,23 @@ func (pl *DynamicResources) PreFilter(ctx context.Context, state *framework.Cycl
 	return nil, nil
 }
 
+func (pl *DynamicResources) validateDeviceClasses(logger klog.Logger, deviceClassName, requestName string) *framework.Status {
+	if deviceClassName == "" {
+		return statusError(logger, fmt.Errorf("request %s: unsupported request type", requestName))
+	}
+
+	_, err := pl.draManager.DeviceClasses().Get(deviceClassName)
+	if err != nil {
+		// If the class cannot be retrieved, allocation cannot proceed.
+		if apierrors.IsNotFound(err) {
+			// Here we mark the pod as "unschedulable", so it'll sleep in
+			// the unscheduleable queue until a DeviceClass event occurs.
+			return statusUnschedulable(logger, fmt.Sprintf("request %s: device class %s does not exist", requestName, deviceClassName))
+		}
+	}
+	return nil
+}
```

Comment:
Unit tests for these new scenarios would be good. What gets allocated is less important (covered by unit tests in the staging repo). What we need here is:
- rejecting a claim with FirstAvailableOf and missing device class when the feature is on
- Was "device class not set and not using FirstAvailableOff" already covered? If not, please also add it (gap in existing test coverage).